drivers/stepper/step_dir_stepper_common.c

```diff
@@ -0,0 +1,263 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Fabian Blatz <fabianblatz@gmail.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "step_dir_stepper_common.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(step_dir_stepper, CONFIG_STEPPER_LOG_LEVEL);
+
+static inline int step_dir_stepper_perform_step(const struct device *dev)
+{
+	const struct step_dir_stepper_common_config *config = dev->config;
+	struct step_dir_stepper_common_data *data = dev->data;
+	int ret;
+
+	switch (data->direction) {
+	case STEPPER_DIRECTION_POSITIVE:
+		ret = gpio_pin_set_dt(&config->dir_pin, 1);
+		break;
+	case STEPPER_DIRECTION_NEGATIVE:
+		ret = gpio_pin_set_dt(&config->dir_pin, 0);
+		break;
+	default:
+		LOG_ERR("Unsupported direction");
+		return -ENOTSUP;
+	}
+	if (ret < 0) {
+		LOG_ERR("Failed to set direction");
+		return ret;
+	}
+
+	if (config->dual_edge) {
+		ret = gpio_pin_toggle_dt(&config->step_pin);
+	} else {
+		ret = gpio_pin_toggle_dt(&config->step_pin);
+		ret |= gpio_pin_toggle_dt(&config->step_pin);
+	}
+
+	if (ret < 0) {
+		LOG_ERR("Failed to toggle step pin");
+		return ret;
+	}
```

Comment:
not a fan of those `ret |=` patterns, there's usually no value in proceeding after the first error, the system is likely broken beyond repair anyway, plus if ret is an error code and for whatever reasons you get two different ones you'd end up returning noise, just return the error at the first fail and move on, how about

```c
ret = gpio_pin_toggle_dt(&config->step_pin);
if (ret < 0) {
        LOG_ERR("Failed to toggle step pin: %d", ret);
        return ret;
}

if (!config->dual_edge) {
        ret = gpio_pin_toggle_dt(&config->step_pin);
        if (ret < 0) {
                LOG_ERR("Failed to toggle step pin: %d", ret);
                return ret;
        }
}
```

Reply:
Very nice idea of simplifying 👍. Will apply.

Reply:
Done